File: src/calc_pi/solution/pi_openmp.f90
program pi_omp
  implicit none
  integer,parameter :: n=1000000
  real*8,parameter :: r=1.0
  integer i,sum
  real*8 x,y,d,pi
  sum=0
  !$omp parallel do private(i,d,x,y) 
  do i=1,n
     call random_number(x)
     call random_number(y)
     d=(2*x*r-r)**2+(2*y*r-r)**2
     if (d.lt.r**2) sum=sum+1
  enddo
  !$omp end parallel do
  pi=4*float(sum)/float(n)
  print *,'The value of pi is',pi
end program pi_omp
  

File: src/calc_pi/solution/pi_openmp.c
#include <omp.h>
#include <math.h>
#include <stdio.h>
int main() {
  int N=1000000;
  double x,y,d;
  double pi,r=1.0;
  int i,sum=0;
#pragma omp parallel for private(i,d,x,y) 
  for (i=0;i<N;i++) {
    x = (double)(rand())/((double)(RAND_MAX)+(double)(1));
    y = (double)(rand())/((double)(RAND_MAX)+(double)(1));
    d = pow(2.*r*x-r,2)+pow(2.*r*y-r,2);
    if (d<pow(r,2)) sum++;
  }
  pi = 4.*(double)(sum)/(double)(N);
  printf("The value of pi is %f\n",pi);
}
  
